diff --git a/log.c b/log.c
index 12db8ca..0174fc2 100644
--- a/log.c
+++ b/log.c
@@ -66,16 +66,23 @@ initlog(int dev)
 
 // Copy committed blocks from log to their home location
 static void 
-install_trans(void)
+install_trans(int trans_to_mem)
 {
   int tail;
 
   for (tail = 0; tail < log.lh.n; tail++) {
-    struct buf *lbuf = bread(log.dev, log.start+tail+1); // read log block
     struct buf *dbuf = bread(log.dev, log.lh.block[tail]); // read dst
-    memmove(dbuf->data, lbuf->data, BSIZE);  // copy block to dst
+    struct buf *lbuf;
+    
+    if (trans_to_mem) {
+        lbuf = bread(log.dev, log.start+tail+1); // read log block
+        memmove(dbuf->data, lbuf->data, BSIZE);  // copy block to dst
+    }
+    
     bwrite(dbuf);  // write dst to disk
-    brelse(lbuf); 
+    if (trans_to_mem) {
+        brelse(lbuf); 
+    }
     brelse(dbuf);
   }
 }
@@ -115,7 +122,7 @@ static void
 recover_from_log(void)
 {
   read_head();      
-  install_trans(); // if committed, copy from log to disk
+  install_trans(1); // if committed, copy from log to disk
   log.lh.n = 0;
   write_head(); // clear the log
 }
@@ -192,7 +199,7 @@ commit()
   if (log.lh.n > 0) {
     write_log();     // Write modified blocks from cache to log
     write_head();    // Write header to disk -- the real commit
-    install_trans(); // Now install writes to home locations
+    install_trans(0); // Now install writes to home locations
     log.lh.n = 0; 
     write_head();    // Erase the transaction from the log
   }
